Add support for transforms of `MultivariateNormal`

sgreenbury opened the pull request:
Closes #579.

sgreenbury commented on autoemulate/experimental/emulators/ensemble.py:
```diff
@@ -111,13 +112,13 @@ def _predict(self, x: Tensor) -> GaussianLike:
-        eye = torch.eye(d, device=sigma_epi.device)  # (dim, dim)
-        jitter_mat = eye.unsqueeze(0).expand(b, d, d) * self.jitter
-        sigma_epi += jitter_mat
+        # b, d = mu_ens.shape
+        # eye = torch.eye(d, device=sigma_epi.device)  # (dim, dim)
+        # jitter_mat = eye.unsqueeze(0).expand(b, d, d) * self.jitter
+        # sigma_epi += jitter_mat
```
I think this can be removed?
```suggestion
```

sgreenbury replied:
Removed

sgreenbury commented on autoemulate/experimental/exploratory/transforms.ipynb:
I think changes to this notebook can be reverted?

cisprague reviewed (commented):
Overall, it looks good. Just some minor comments:
1. Perhaps use `broadcast_to` or `expand` rather than `repeat` to avoid materializing large matrices.
2. Perhaps use `@torch.inference_mode()` for computations that don't require end-to-end functionality.

sgreenbury commented on autoemulate/experimental/transforms/base.py:
```diff
@@ -264,22 +295,29 @@ def _inverse_sample_gaussian_like(
+
+    # Compute the mean and covariance of the samples
     mean = samples.mean(dim=0)
+    if not full_covariance:
```
This would be great to handle more than just either full or diag - perhaps we can revisit this after #471 ?

sgreenbury commented on autoemulate/experimental/transforms/vae.py:
```diff
@@ -143,11 +143,29 @@ def log_abs_det_jacobian(self, x, y):
+        n = y.shape[0]
+
+        # Reshape jacobian for shape of cov_y (n_tasks x n_samples)
+        return jacobian.reshape(n * jacobian.shape[1], -1)
```
See https://github.com/alan-turing-institute/autoemulate/pull/586#discussion_r2190108986

sgreenbury commented on autoemulate/experimental/transforms/utils.py:
```diff
@@ -45,31 +47,66 @@ def make_positive_definite(
+            cov = (cov + cov.transpose(-1, -2)) / 2
+            min_eigval *= 10
+
+    # Spectral approach by clamping minimum and maximum eigenvalue
```
The second block repeats the above but also clamps the eigenvalue to a maximum (since it failed when only clamping to 0).

sgreenbury commented:
Thanks @cisprague for the review and comments - I've:
- Addressed some mentioned above, some unfortunately such as replacing replacing expand raised exception
- Opened #601 to make the improvements discussed for `make_positive_definite` and capture performance improvements
- Adjusted the ensembles to only have jitter applied to the covariance matrices within the model

sgreenbury commented:
This has been updated further in https://github.com/alan-turing-institute/autoemulate/pull/586/commits/729d58f3dfb1a1a517e1d2c8c56b33a6f5748665 so that:
- API is clearer for `make_positive_definite`
- The heuristic for eigenvalue clamping is based on conditioning number and set to `False` where possible

@cisprague - are you happy to merge this with any further revisions in #601?
